src/firestarter/DumpRegisterWorker.cpp

```diff
@@ -0,0 +1,186 @@
+    dist += val & 1;
+  }
+
+  return dist;
+}
+
+static std::string registerNameBySize(unsigned registerSize) {
+  switch (registerSize) {
+  case 2:
+    return "xmm";
+  case 4:
+    return "ymm";
+  case 8:
+    return "zmm";
+  default:
+    return "unknown";
+  }
+}
+} // namespace
+
+int Firestarter::initDumpRegisterWorker(std::chrono::seconds dumpTimeDelta,
+                                        std::string dumpFilePath) {
+
+  auto data = new DumpRegisterWorkerData(this->loadThreads.begin()->second,
+                                         dumpTimeDelta, dumpFilePath);
+
+  pthread_create(&this->dumpRegisterWorkerThread, NULL, dumpRegisterWorker,
+                 std::ref(data));
+
+  return EXIT_SUCCESS;
+}
+
+void Firestarter::joinDumpRegisterWorker(void) {
+  pthread_join(this->dumpRegisterWorkerThread, NULL);
+}
+
+void *Firestarter::dumpRegisterWorker(void *dumpRegisterData) {
+
+  pthread_setname_np(pthread_self(), "DumpRegWorker");
+
+  auto data = reinterpret_cast<DumpRegisterWorkerData *>(dumpRegisterData);
+
+  auto payload = data->loadWorkerData->config->payload;
+  int registerCount = payload->registerCount;
+  int registerSize = payload->registerSize;
+  std::string registerPrefix = registerNameBySize(registerSize);
+  auto offset = sizeof(DumpRegisterStruct) / sizeof(unsigned long long);
+
+  auto dumpRegisterStruct = reinterpret_cast<DumpRegisterStruct *>(
+      data->loadWorkerData->addrMem - offset);
+
+  auto dumpVar = reinterpret_cast<volatile unsigned long long *>(
+      &dumpRegisterStruct->dumpVar);
+  // memory of simd variables is before the padding
+  volatile unsigned long long *dumpMemAddr =
+      dumpRegisterStruct->padding - registerCount * registerSize;
+
+  // TODO: handle error
+  // TODO: maybe use aligned_malloc to make memcpy more efficient and don't
+  // interrupt the workload as much?
+  unsigned long long *last = reinterpret_cast<unsigned long long *>(
+      malloc(sizeof(unsigned long long) * offset));
+  unsigned long long *current = reinterpret_cast<unsigned long long *>(
+      malloc(sizeof(unsigned long long) * offset));
+
+  auto dumpFile = std::ofstream(data->dumpFilePath / "hamming_distance.csv");
+
+  // dump the header to the csv file
+  dumpFile << "total_hamming_distance,";
+  for (int i = 0; i < registerCount; i++) {
+    for (int j = 0; j < registerSize; j++) {
+      dumpFile << registerPrefix << i << "[" << j << "]";
+
+      if (j != registerSize - 1) {
+        dumpFile << ",";
+      }
+    }
+
+    if (i != registerCount - 1) {
+      dumpFile << ",";
+    }
+  }
+  dumpFile << std::endl << std::flush;
+
+  // do not output the hamming distance for the first run
+  bool skipFirst = true;
+
+  // continue until stop and dump the registers every data->dumpTimeDelta
+  // seconds
+  for (; *data->loadWorkerData->addrHigh != LOAD_STOP;) {
+    // signal the thread to dump its largest SIMD registers
+    *dumpVar = DumpVariable::Start;
+    __asm__ __volatile__("mfence;");
```

Comment:
Yes this would be generally what i would like to use, but as this variable gets polled by the load generating threads in assembly; they should not block during synchronisation. This is the variant used in previous versions of FIRESTARTER for load-level synchronisation and is known to work. I would also like this to be architecture-agnostic, but haven't had time to search for a better solution yet.